### PR TITLE
kv/rocksdb: supports SliceParts interface

### PR DESCRIPTION
**Pull request by yuyuyu101**

### CODE REVIEW

File: src/kv/RocksDBStore.cc [hunk 1]
```diff
@@ -577,10 +577,10 @@ void RocksDBStore::RocksDBTransactionImpl::set(
...
 			    to_set_bl.length()));
   } else {
-    // make a copy
-    bufferlist val = to_set_bl;
-    bat.Put(rocksdb::Slice(key),
-	     rocksdb::Slice(val.c_str(), val.length()));
+    rocksdb::Slice key_slice(key);
+    rocksdb::Slice value_slices[to_set_bl.buffers().size()];
+    bat.Put(nullptr, rocksdb::SliceParts(&key_slice, 1),
+            rocksdb::SlicePartsi(&value_slices, to_set_bl.buffers().size()));
```
Comment (liewegas) [inline src/kv/RocksDBStore.cc]: I don't see the part where value_slices is populated with the bufferptr data pointers...

**Comment by yuyuyu101** [comment]
oh, yes... lost git operation

**Comment by yuyuyu101** [comment]
@liewegas @markhpc the third commit add prefix_filter supports in rocksdb(https://github.com/facebook/rocksdb/wiki/Prefix-Seek-API-Changes). it will help a lot for million keys case. But it will lose iterate backward capacity. I scan the iterate use case, if accepted this, it won't be problem for current omap iterator users.

File: src/kv/RocksDBStore.cc [hunk 2]
```diff
@@ -34,6 +34,16 @@ using std::string;
...
 #define dout_prefix *_dout << "rocksdb: "
 
+static rocksdb::SliceParts prepare_sliceparts(bufferlist &bl, rocksdb::Slice *slices)
+{
+  unsigned n = 0;
+  for (std::list<buffer::ptr>::const_iterator p = _buffers.begin();
+       p != _buffers.end(); ++p, ++n) {
+    slices[n].data_ = (void *)p->c_str();
+    slices[n].size_ = p->length();
+  }
```
Comment (liewegas) [inline src/kv/RocksDBStore.cc]: missing return value?

**Comment by liewegas** [comment]
Slice parts thing looks great (once it compiles :).

I can't think of any case where we use a reverse iterator.  I'm not clear where we want to use the prefix, though.. are you thinking for the bluestore omap key prefix for each object?  Or for the bluestore one-character type prefixes (it seems unlikely to help there?).

**Comment by yuyuyu101** [comment]
@liewegas yes, this commit just is a example.

we could define "PREFIX"(1 byte) + "INDEX_PREFIX"(8 bytes) as the rocksdb prefix
I expect we could see benefit from bluestore onode shards, rgw bucket index lookup/scan.

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-05-29_22:34:44-rados-wip-sage-testing---basic-smithi/

**Comment by wjwithagen** [comment]
@yuyuyu101 @liewegas 
Not sure why it only surfaces now, but Clang does not like the slices, or the way they are used.
This is on a freshly installed system, so could also be missing something.
```
/usr/srcs/Ceph/work.v20170726/ceph/src/kv/RocksDBStore.cc:611:32: error: variable length array of non-POD element type
      'rocksdb::Slice'
    rocksdb::Slice value_slices[to_set_bl.buffers().size()];
                               ^
/usr/srcs/Ceph/work.v20170726/ceph/src/kv/RocksDBStore.cc:632:32: error: variable length array of non-POD element type
      'rocksdb::Slice'
    rocksdb::Slice value_slices[to_set_bl.buffers().size()];
                               ^
/usr/srcs/Ceph/work.v20170726/ceph/src/kv/RocksDBStore.cc:710:32: error: variable length array of non-POD element type
      'rocksdb::Slice'
    rocksdb::Slice value_slices[to_set_bl.buffers().size()];
                               ^
```

**Comment by tchaikov** [comment]
@wjwithagen could use a `vector<rocksdb::Slice>` to workaround this.

**Comment by tchaikov** [comment]
@wjwithagen see #16615

**Pull request merged**